pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,430 @@
+				klog.ErrorS(err, "Failed to parse ICMP message")
+				continue
+			}
+			if msg.Type != ipv6.ICMPTypeEchoReply {
+				klog.V(4).InfoS("Failed to match ICMPTypeEchoReply")
+				continue
+			}
+		}
+
+		echo, ok := msg.Body.(*icmp.Echo)
+		if !ok {
+			klog.ErrorS(nil, "Failed to assert type as *icmp.Echo")
+			continue
+		}
+
+		klog.V(4).InfoS("Recv ICMP message", "IP", destIP, "echo", echo)
+
+		// Parse the time from the ICMP data
+		sentTime, err := time.Parse(time.RFC3339Nano, string(echo.Data))
+		if err != nil {
+			klog.ErrorS(err, "Failed to parse time from ICMP data")
+			continue
+		}
+
+		// Calculate the round-trip time
+		end := time.Now()
+		rtt := end.Sub(sentTime)
+
+		// Update the latency store
+		mutator := func(entry *NodeIPLatencyEntry) {
+			entry.LastSendTime = sentTime
+			entry.LastRecvTime = end
+			entry.LastMeasuredRTT = rtt
+		}
+		m.latencyStore.SetNodeIPLatencyEntry(destIP, mutator)
+	}
+}
+
+// pingAll sends ICMP messages to all the Nodes.
+func (m *NodeLatencyMonitor) pingAll(ipv4Socket, ipv6Socket net.PacketConn) {
+	nodeIPs := m.latencyStore.ListNodeIPs()
+	for _, toIP := range nodeIPs {
+		if toIP.To4() != nil && ipv4Socket != nil {
+			if err := m.sendPing(ipv4Socket, toIP); err != nil {
+				klog.ErrorS(err, "Failed to send ICMP message to Node", "IP", toIP)
+			}
+		} else if toIP.To16() != nil && ipv6Socket != nil {
+			if err := m.sendPing(ipv6Socket, toIP); err != nil {
+				klog.ErrorS(err, "Failed to send ICMP message to Node", "IP", toIP)
+			}
+		} else {
+			klog.ErrorS(nil, "Failed to send ICMP message to Node")
+		}
+	}
+}
+
+// Run starts the NodeLatencyMonitor.
+func (m *NodeLatencyMonitor) Run(stopCh <-chan struct{}) {
+	go m.nodeLatencyMonitorInformer.Informer().Run(stopCh)
+	go m.nodeInformer.Informer().Run(stopCh)
+	go m.monitorLoop(stopCh)
+
+	<-stopCh
+}
+
+// monitorLoop is the main loop to monitor the latency of the Node.
+func (m *NodeLatencyMonitor) monitorLoop(stopCh <-chan struct{}) {
+	// Low level goroutine to handle ping loop
+	var ticker *time.Ticker
+	var tickerCh <-chan time.Time
+	var ipv4Socket, ipv6Socket net.PacketConn
+	var err error
+
+	defer func() {
+		if ipv4Socket != nil {
+			ipv4Socket.Close()
+		}
+		if ipv6Socket != nil {
+			ipv6Socket.Close()
+		}
+		if ticker != nil {
+			ticker.Stop()
+		}
+	}()
+
+	// Update current ticker based on the latencyConfig
+	updateTicker := func(interval time.Duration) {
+		if ticker != nil {
+			ticker.Stop() // Stop the current ticker
+		}
+		ticker = time.NewTicker(interval)
+		tickerCh = ticker.C
+	}
+
+	// Start the pingAll goroutine
+	for {
+		select {
+		case <-tickerCh:
+			// Try to send pingAll signal
+			m.pingAll(ipv4Socket, ipv6Socket)
+		case <-stopCh:
+			// Close current sockets, the recvPing will stop by itself
+			if ipv4Socket != nil {
+				ipv4Socket.Close()
+			}
+			if ipv6Socket != nil {
+				ipv6Socket.Close()
+			}
+			// Stop the ticker loop
+			if ticker != nil {
+				ticker.Stop()
+			}
```

Comment:
this is redundant: you already have a `defer` function that does the exact same thing. You should only have one or the other.

Reply:
Thanks, I have removed them.